Revised d_solveprep to prohibit wind builds in LA

ssattler opened the pull request:
Added code to prohibit wind builds in LA until after 2032, per Sam Gomberg and Paul Arbaje's review of near-term builds in LA. Code originally written in 2020 with help from Wesley Cole. Originally these lines were added in d_solveprep after this section: 
recperc(rpscat,st,t) = round(recperc(rpscat,st,t),4) ; 
rsc_fin_mult(i,r,t) = round(rsc_fin_mult(i,r,t),3) ; 
tranloss(r,rr,trtype) = round(tranloss(r,rr,trtype),4) ; 

rsc_fin_mult and tranloss are no longer located in d_solveprep in the 2024 version of ReEDS. I added the new lines after recperc, which is now in the "round parameters" section of d_solveprep. 

## Summary

## Technical Details
### Implementation notes

### Additional changes (if any)

### Switches added/removed/changed (if any)

### Issues resolved (if any)

### Known incompatibilities (if any)

### Relevant sources or documentation (if any)

### Is the model cleaner/better/faster/smaller than before? If so, how?

## Validation / Comparison Report
### Link to report and/or example plots

### (if pertinent) How big is the default run folder (runs/{base}_ref_seq/) before and after the change?

### (if pertinent) What is the run time for ref_seq on the same machine before and after the change?

ssattler committed:
Revised d_solveprep to prohibit wind builds in LA

Added code to prohibit wind builds in LA until after 2032. Code originally written in 2020 with help from Wesley Cole. Originally this added after these lines in d_solveprep:
recperc(rpscat,st,t) = round(recperc(rpscat,st,t),4) ;
rsc_fin_mult(i,r,t) = round(rsc_fin_mult(i,r,t),3) ;
tranloss(r,rr,trtype) = round(tranloss(r,rr,trtype),4) ;
rsc_fin_mult and tranloss are no longer located in d_solveprep. I added the new lines after recperc, which is now in the "round parameters" section.

samgdotson reviewed (changes requested):
Thanks @ssattler, I just have a few questions that need to be addressed before we can merge this. Also, I will create a new branch called `bau-scenario` and I will try to edit the PR so that it gets merged into that branch, rather than `main`, since this is a model-wide change that we're only going to use for one scenario (right?).

ssattler commented on d_solveprep.gms:
```diff
@@ -102,6 +102,11 @@ trans_cost_cap_fin_mult_noITC(t) = round(trans_cost_cap_fin_mult_noITC(t),3) ;
 
+* Restrict UPV in LA, per PA and SG recommendation - SSS 4/9/25
+Set t_notallowed(t) / 2020*2032 / ;
+m_required_prescriptions("upv",r,t_notallowed)$r_st(r,"LA") = 0 ;
```
You're absolutely correct, this needs to be changed from upv to wind-ons. I am really glad for this catch because this raises an additional issue, onshore wind is built in resource regions, not PCAs. So I'm going to look at this closer to see if your correction above works, or if it needs to be adjusted further.

ssattler commented on d_solveprep.gms:
```diff
@@ -102,6 +102,11 @@ trans_cost_cap_fin_mult_noITC(t) = round(trans_cost_cap_fin_mult_noITC(t),3) ;
+* Restrict UPV in LA, per PA and SG recommendation - SSS 4/9/25
+Set t_notallowed(t) / 2020*2032 / ;
+m_required_prescriptions("upv",r,t_notallowed)$r_st(r,"LA") = 0 ;
+INV.fx(i,v,r,t_notallowed)$[PV(i)$ r_st(r,'LA')] = 0 ;
```
See above, Wesley Cole wrote that line when we were restricting solar in MD in 2020. But again, wind-ons is at the resource region ('s') level and not the PCA level, so I want to dig a little and will get back to you.

ssattler committed:
revise d_solveprep to restrict wind builds in LA before 2032

added code to restrict wind builds in LA in d_solveprep
Wesley Cole advised to also include restricting the INV.fx (an investment variable found in d2_varfix) from investing in wind in LA for the same time period.

samgdotson committed:
Merge pull request #13 from ucsusa/ssattler-patch-1

revise d_solveprep to restrict wind builds in LA before 2032

samgdotson commented:
I tried running a test case with this and got the following error:

```
--- . d_solveprep.gms(107) 79 Mb 1 Error
*** Error 170 in /home/ce-admin/research/ReEDS-2.0/runs/wind_limit_EER_Current_LowDC_UCS/d_solveprep.gms
    Domain violation for element
--- . d_solveprep.gms(108) 79 Mb 2 Errors
*** Error 170 in /home/ce-admin/research/ReEDS-2.0/runs/wind_limit_EER_Current_LowDC_UCS/d_solveprep.gms
    Domain violation for element
--- . d_solveprep.gms(259) 79 Mb
--- createmodel.gms(4) 79 Mb 2 Errors
*** Status: Compilation error(s)
--- Job createmodel.gms Stop 04/17/25 18:21:26 elapsed 0:00:01.467
```

The reference to lines 107 and 108 make sense because those are a couple of the new lines....

ssattler commented:
Error 170 means that an element does not belong to the set. I'll think about this.

samgdotson commented:
Here’s what I’ve gleaned so far. There are sets called “onswind” and “ofswind” and the members of these sets are technology types “wind-ons_{class}” where class is the wind resource class. 

What I think is happening is that making “onswind” a string is referencing a specific technology type called “onswind,” which I don’t think exists. 

So far, I tried replacing that with “wind-ons” and got a different error that could be caused by something else.

ssattler commented:
What's the error? I can try to run it down.

samgdotson commented:
I tried using `onswind` and `ofswind` (without quotes) to reference the sets directly, and got the following error
```bash
--- . c_supplyobjective.gms(361) 79 Mb
--- createmodel.gms(4) 79 Mb
--- . d_solveprep.gms(107) 79 Mb 1 Error
*** Error 171 in /home/ce-admin/research/ReEDS-2.0/runs/wind-limit-t3_EER_Current_LowDC_UCS/d_solveprep.gms
    Domain violation for set
--- . d_solveprep.gms(108) 79 Mb 2 Errors
*** Error 171 in /home/ce-admin/research/ReEDS-2.0/runs/wind-limit-t3_EER_Current_LowDC_UCS/d_solveprep.gms
    Domain violation for set
--- . d_solveprep.gms(259) 79 Mb
--- createmodel.gms(4) 79 Mb 2 Errors
*** Status: Compilation error(s)
--- Job createmodel.gms Stop 04/18/25 15:51:06 elapsed 0:00:01.499
missing g00files/wind-limit-t3_EER_Current_LowDC_UCS.g*
wind-limit-t3_EER_Current_LowDC_UCS has finished
```

samgdotson commented:
Here's my traceback...

`m_required_prescriptions` is defined in `b_inputs.gms` with 

```gams
# line 2234 of b_inputs.gms
Parameter m_required_prescriptions(pcat, r, t)
```

`pcat` is a set defined at runtime when the input files are generated via `copy_files.py`. After that, there is a new file called `pcat.csv` with the following data (for example...)

```
*seldom-used pound/hashtag populates
*elements of this set with the indicated set
#i
upv
dupv
pvb
wind-ons
wind-ofs
csp-ws
geohydro_allkm
egs_allkm
egs_nearfield
```

From this I'm concluding the following:
1. "onswind" doesn't work because it is the name of a set AND is not a member of the set `pcat`. 
2. `onswind` doesn't work because it is a different set from `pcat` which includes more technologies that I mentioned earlier, none of which are actually listed in `pcat`. 
3. I'm surprised that using "wind-ons" _didn't_ work -- so I'm running it again. Strangely/fortunately, it got a lot farther than the first try. Which is super confusing. I'll keep you updated.

samgdotson commented:
OK, good news -- the model was created successfully and ran for awhile. But there's a new error.

```bash

--- Conflict Refiner status (31): minimal conflict
--- Time spent on conflict find: 8.3 seconds

--- A conflict is a set of equations and variables (i.e. a submodel) which is
--- infeasible but becomes feasible if any one equation or variable bound is dropped.
--- A problem may contain several independent conflicts but only one will be found per run.

Number of equations in conflict: 5
fixed: eq_forceprescription(wind-ons,p58,2023) = 210
lower: eq_refurblim(wind-ons_4,p58,2023) > 0
lower: eq_refurblim(wind-ons_5,p58,2023) > 0
lower: eq_refurblim(wind-ons_6,p58,2023) > 0
lower: eq_refurblim(wind-ons_7,p58,2023) > 0
```

This error tells me we need to figure out the prescribed builds for wind in Louisiana. The limit cannot be zero.

OR we should specify in `inputs/capacity_exogenous/wind-ons_prescribed_builds_{scenario}.csv` that there are no prescribed builds in p58 or p86 (louisiana).

samgdotson commented:
closing with preference for PR #16
